File: J9Ch/src/J_2_Binary_Search/Required_10/_459_Closest_Number_in_Sorted_Array_Easy.cpp
/**459. Closest Number in Sorted Array
 * Easy*/



class Solution {
public:
    /**
     * @param A an integer array sorted in ascending order
     * @param target an integer
     * @return an integer
     */
    int closestNumber(vector<int>& A, int target) {
        // Write your code here
        int l = 0, r = A.size() - 1;
        if (r < 0)
            return -1;
        int ans = abs(A[0] - target);
        int index = 0;
        while (l <= r) {
            int mid = (l + r) >> 1;
            if (ans > abs(A[mid] - target)) {
                ans = abs(A[mid] - target);
                index = mid;
            }
            if (A[mid] >= target)
                r = mid - 1;
            else
                l = mid + 1;
        }
        return index;
    }
};

File: J9Ch/src/J_2_Binary_Search/Required_10/_459_Closest_Number_in_Sorted_Array_Easy.py
'''/**459. Closest Number in Sorted Array
 * Easy'''


class Solution:
    # @param {int[]} A an integer array sorted in ascending order
    # @param {int} target an integer
    # @return {int} an integer
    def closestNumber(self, A, target):
        if not A:
            return -1
        start, end = 0, len(A) - 1
        while start + 1 < end:
            mid = (start + end) / 2
            if A[mid] == target:
                return mid
            elif A[mid] > target:
                end = mid
            else:
                start = mid
        if A[end] - target > target - A[start]:
            return start
        else:
            return end

A = [1,4,6,10,20]
B = 21
sol = Solution()
print(sol.closestNumber(A, B))
print(Solution().closestNumber(A, B))